schema/representations.md

```diff
@@ -0,0 +1,257 @@
+
+Each of these representation strategies will be explored in detail in the
+"Representation Strategy Reference" section, below.
+
+
+Representation Strategy Reference
+---------------------------------
+
+In this section, we'll go into detail on each representation strategy,
+and include examples of both schemas using it and data matching it.
+See the "Kinds and their Representations" section for an introduction to
+representation strategies.
+
+
+### struct map representation
+
+(TODO mostly interesting because defaults)
+(which maybe we should discuss more in the Kinds page?  belongs with nullable, optional, and other cardinality talk.)
+
+### struct tuple representation
+
+Tuple representation allows structs to be packed into a list representation.
+This list representation drops the struct fieldnames from the representation
+entirely and relies purely upon the order of elements to denote meaning.
+As such, it's generally a very compact way to represent data.
+
+Tuple representations should be used cautiously.  Since they contain little
+in the way of "self-describing" information, tuple representations can make for
+very fragile protocols, increase the difficulty of migrations, and make
+serialized data incomprehensible without the schema information in hand.
+
+### struct map representation example schema with defaults
+
+```ipldsch
+type TypeList struct {
+	valueType TypeTerm
+	valueNullable Bool
+} representation map {
+	field valueNullable default "false"
+}
+```
+
+#### struct tuple representation example schema
+
+```ipldsch
+```
+
+#### struct tuple representation example data
+
+```json
+```
+
+### struct stringjoin representation example schema
+
+```ipldsch
+## Fizzlebop is a pair of fields which serializes as "part1:part2" as a string.
+type Fizzlebop struct {
+	a String
+	b String
+} representation stringjoin {
+	join ":"
+}
+```
+
+### map stringpairs representation example
+
+Say we're doing something awfully like the mount options in an /etc/fstab file:
+
+```ipldsch
+## MountOptions serializes as a string of comma-separated "k=v" entries, e.g. "k1=v1,k2=v2".
+## This serial representation is limited (n.b. The domain of keys must exclude "=" and values must exclude ","!),
+## but we've chosen it because other parts of our application protocol are drastically simplified by limiting this to a string.
+type MountOptions map {String:String} representation stringpairs {
+	innerDelim "="
+	entryDelim ","
+}
+```
+
+### union keyed representation example
+
+```ipldsch
+type MyKeyedUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation keyed
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKeyedUnion` (shown as JSON) is:
+
+```json
+{"foo": {"froz": true}}
+```
+
+This data would also match, as the other type:
+
+```json
+{"bar": 12}
+```
+
+### union kinded representation example
+
+```ipldsch
+type MyKindedUnion union {
+	| Foo map
+	| Bar int
+} representation kinded
```

Comment:
I think we should drop the kinds from kinded unions and infer them from the info we already have on those types (we know `Foo` is a map and we know `Bar` is an int). The parsing pattern is broken when kinded unions have unquoted kinds in each element descriptor but the other union types have quotes strings for the various discriminators. This is going to make learning more difficult, particularly for example-driven folks who will keep on seeing weird discrepancies.

I agree that moving the discriminators down into the `representation` block would become tedious very fast, requiring a doubling up of each element. But having it up in the main block undermines your original ideal of keeping representation-specific information out of the main block. Maybe if we're doing a special case for unions we should consider how much of the representation block can be removed. A type can only be of one representation anyway, it's not as the main block needs to be kept clean for multiple representation forms, they're already tied together.

I can at least see any of the `field` descriptors in the representation needing to be lifted out to the main block by this same logic. And while at it, I'd put the name _after_ the type to bring them closer to unions. I'd even be tempted to quote them for the same reasons.

```
type TypeList struct {
	foo Foo
	bar Bool
	baz Int
} representation map {
	field foo optional
	field bar nullable default "false"
	field baz optional nullable
}
```

->

```
type TypeList struct {
	Foo "foo" optional
	Bool "bar" nullable omitted "false"
	Int "baz" optional nullable
} representation map
```

as a tuple:

```
type TypeList struct {
	Foo "foo"
	Bool "bar" nullable
	Int "baz" optional nullable
} representation tuple
```

Maybe the rule is that anything in the `representation` block should be required generally for the representation (discriminantKey etc.) rather than specifically for individual fields. Anything needed for the fields should go in the main block coupled with the fields.

The parser will then still have to backtrack and error if it finds field descriptors with too much, or the wrong kind of information. Front-loading a representation block doesn't seem to work (too messy), but just collapsing it in might:

```
type MyEnvelopeUnion union {
	| Foo "foo"
	| Bar "bar"
} representation envelope {
 	discriminantKey "tag"
 	contentKey "msg"
}
```

->

```
type MyEnvelopeUnion union envelope {
 	$discriminantKey "tag"
 	$contentKey "msg"
	| Foo "foo"
	| Bar "bar"
}
```

That's fantastic for parsing, but maybe I making too much of a dogs breakfast of your vision?

Reply:
OK, maybe scrap that last bit, I'm reflecting on its ugliness. I'd still like to explore the fields problem though because the representation is already mixed up in the definition and we're sitting in an awkward middle that I think we can move out of.

Reply:
Not a fan of dropping kinds from the declaration syntax of kinded unions -- it *is* redundant, for sure, but imo it's going to be a useful redundance, and it's a useful lexically local clarity of intent (i.e. I wanna *see* it in the union defn while I'm typing it out if I accidentally use two things that are a map) and it makes for clearer error reporting (i.e. if I type out a lie there that doesn't match the member type defn's kind, it's very clear exactly which lines to point at as the error, rather than having to point at a whole range of lines and leave the user to figure out which one is conflictory.)

To the rest of it ... and per conversation in other channels... Yeah, I'm super willing to entertain (if not already full on convinced) there's some weird bits here, and maybe we could reach a better and more internally consistent thing.

Getting there is gonna be bigger than fits on the discussion of one PR though; let's land this one, call it a checkpoint, and open more, yeh?

Reply:
Just to toss this out there, one variation in syntax I've been trying out is:

```
type SomeStruct struct {
    fieldname nullable OtherType (alias "foo")
}
```

The rule here would be "things that would have gone in the representation block, but are specific to a field... go after the field, but in parens".  (Thus, modifiers like "nullable" still stay outside the parens -- they're still a cardinality changer, so they're still clearly a type-defn thing.)

But this is just another idea.  And not the one in this PR.  Just airing it here as an idea of a direction we could examine further in a future PR.